### PR TITLE
Verify admin email address

### PR DESCRIPTION
**Pull request by rauchy**
verify

This PR adds support for verifying e-mail addresses. Invited users are automatically verified.

### CODE REVIEW

File: redash/handlers/users.py [hunk 1]
```diff
@@ -119,15 +118,9 @@ def post(self):
             'object_type': 'user'
         })
 
-        if request.args.get('no_invite') is not None:
```
Comment (rauchy) [inline redash/handlers/users.py]: TODO: bring back no_invite

File: redash/authentication/account.py [hunk 2]
```diff
@@ -35,6 +42,15 @@ def validate_token(token):
     return serializer.loads(token, max_age=max_token_age)
 
 
+def send_verify_email(user, org):
+    context = dict(user=user, verify_url=verify_link_for_user(user))
```
Comment (jezdez) [inline redash/authentication/account.py]:
```suggestion
    context = {
        'user': user,
        'verify_url': verify_link_for_user(user),
    }
```

File: redash/handlers/authentication.py [hunk 3]
```diff
@@ -102,6 +122,14 @@ def forgot_password(org_slug=None):
     return render_template("forgot.html", submitted=submitted)
 
 
+@routes.route(org_scoped_rule('/send_verification'))
+def send_verification(org_slug=None):
+    if not current_user.is_email_verified:
+        send_verify_email(current_user, current_org)
+
+    return "Please check your e-mail inbox in order to verify your address.", 200
```
Comment (jezdez) [inline redash/handlers/authentication.py]:
```suggestion
    return "Please check your email inbox in order to verify your email address.", 200
```

File: redash/templates/emails/verify.html [hunk 4]
```diff
@@ -0,0 +1,15 @@
+{% extends "emails/layout.html" %}
+
+{% block content %}
+
+<p class="intercom-align-left" style="line-height: 1.5; margin: 0 0 17px; text-align: left !important" align="left">Hi {{ user.name }},</p>
+<h2 class="intercom-align-left" style="color: #282F33; font-size: 18px; font-weight: bold; margin-bottom: 7px; margin-top: 30px; text-align: left !important" align="left">
+    Please verify that {{ user.email }} is your correct e-mail address by visiting the following link:
```
Comment (jezdez) [inline redash/templates/emails/verify.html]:
```suggestion
    Please verify that {{ user.email }} is your correct email address by visiting the following link:
```

File: redash/templates/emails/verify.txt [hunk 5]
```diff
@@ -0,0 +1,7 @@
+Hi {{ user.name }},
+
+Please verify that {{ user.email }} is your correct e-mail address by visiting the following link:
```
Comment (jezdez) [inline redash/templates/emails/verify.txt]:
```suggestion
Please verify that {{ user.email }} is your correct email address by visiting the following link:
```

File: redash/templates/verify.html [hunk 6]
```diff
@@ -0,0 +1,14 @@
+{% extends "layouts/signed_out.html" %}
+{% block title %}E-mail Verified{% endblock %}
+{% block head %}
+<meta http-equiv="refresh" content="5;URL='{{ url_for('redash.index') }}'" />
+{% endblock %}
+{% block content %}
+<div class="fixed-width-page">
+  <div class="bg-white tiled">
+      <div>
+          Thanks for verifying your e-mail address. You will be redirected back to the app within a few seconds.
```
Comment (jezdez) [inline redash/templates/verify.html]:
```suggestion
          Thanks for verifying your email address. You will be redirected back to the app within a few seconds.
```

File: redash/templates/verify.html [hunk 7]
```diff
@@ -0,0 +1,14 @@
...
+      </div>
```
[the remark this answers is not in the document]
Comment (rauchy) [reply redash/templates/verify.html]: Done. I didn't know you could disable auto-redirects.

File: client/app/pages/users/new.html [hunk 8]
```diff
@@ -25,8 +25,6 @@ <h3>New User</h3>
 
         <div ng-if="user.created" class="alert alert-success alert-invited">
           <h4>The user has been created and should receive an invite email soon</h4>
```
Comment (arikfr) [inline client/app/pages/users/new.html]:
```suggestion
          <h4>The user has been created and should receive an invite email soon.</h4>
```

File: redash/handlers/authentication.py [hunk 9]
```diff
@@ -102,6 +122,14 @@ def forgot_password(org_slug=None):
     return render_template("forgot.html", submitted=submitted)
 
 
+@routes.route(org_scoped_rule('/send_verification'))
+def send_verification(org_slug=None):
+    if not current_user.is_email_verified:
+        send_verify_email(current_user, current_org)
+
+    return "Please check your email inbox in order to verify your email address.", 200
```
Comment (arikfr) [inline redash/handlers/authentication.py]: This should return JSON.

File: redash/handlers/authentication.py [hunk 10]
```diff
@@ -102,6 +122,14 @@ def forgot_password(org_slug=None):
     return render_template("forgot.html", submitted=submitted)
 
 
+@routes.route(org_scoped_rule('/send_verification'))
```
Comment (arikfr) [inline redash/handlers/authentication.py]: Need to limit this only to a POST request.

File: client/app/components/empty-state/empty-state.html [hunk 11]
```diff
@@ -1,3 +1,4 @@
+<div ng-if="!$ctrl.isEmailVerified" class="alert alert-danger">We have sent an email with a confirmation link to your email address. Please follow the link to verify your email address. <a ng-click="verifyEmail()">Resend email</a>.</div>
```
Comment (arikfr) [inline client/app/components/empty-state/empty-state.html]: We're showing the EmptyState component in multiple places. Was it intentional that you used it?

Comment (arikfr) [reply client/app/components/empty-state/empty-state.html]: Also, maybe a bit of a less scary color for the message?

@ranbena

File: client/app/components/empty-state/empty-state.js [hunk 12]
```diff
@@ -45,6 +46,12 @@ const EmptyStateComponent = {
         },
       });
     };
+
+    $scope.verifyEmail = () => {
```
Comment (arikfr) [inline client/app/components/empty-state/empty-state.js]: You don't need `$scope` here. Just put a method on `this` and in the template, use `$ctrl.verifyEmail`.

File: redash/handlers/authentication.py (hunk 10 above)
Comment (rauchy) [reply redash/handlers/authentication.py]: Already done earlier, probably didn't push yet.

File: client/app/components/empty-state/empty-state.js (hunk 12 above)
Comment (rauchy) [reply client/app/components/empty-state/empty-state.js]: image

File: client/app/components/empty-state/empty-state.html (hunk 11 above)
Comment (rauchy) [reply client/app/components/empty-state/empty-state.html]: re: color - another option is this:

image

Comment (arikfr) [reply client/app/components/empty-state/empty-state.html]: 🤢

Comment (rauchy) [reply client/app/components/empty-state/empty-state.html]: > We're showing the EmptyState component in multiple places. Was it intentional that you used it?

Not really. Moved to `<home-page>`.

**Pull request merged**